src/psyclone/domain/common/extract_driver_creator.py

```diff
@@ -0,0 +1,711 @@
+        # property is used. The name of a derived type is 'flattened', i.e.
+        # all '%' are replaced with '_', and this is then declared as a
+        # non-structured type. We also need to make sure that a flattened
+        # name does not clash with a variable declared by the user. We use
+        # the structured name (with '%') as tag to handle this.
+        for reference in all_references:
+            if not isinstance(reference, StructureReference):
+                continue
+            old_symbol = reference.symbol
+            if old_symbol.datatype.name != "r2d_field":
+                fortran_string = writer(reference)
+                raise InternalError(
+                    "Error when constructing driver for '{0}': "
+                    "Unknown derived type '{1}' in reference '{2}'."
+                    .format(sched.name, old_symbol.datatype.name,
+                            fortran_string))
+            # We have a structure reference to a field, flatten it, and
+            # replace the StructureReference with a new Reference to this
+            # flattened name (e.g. `fld%data` becomes `fld_data`)
+            self.flatten_reference(reference, symbol_table, writer=writer)
+
+    # -------------------------------------------------------------------------
+    @staticmethod
+    def add_call(program, name, args):
+        '''This function creates a call to the subroutine of the given name,
+        providing the arguments. The call will be added to the program and
+        to the symbol table.
+
+        :param program: the PSyIR Routine to which any code must \
+            be added. It also contains the symbol table to be used.
+        :type program: :py:class:`psyclone.psyir.nodes.Routine`
+        :param str name: name of the subroutine to call.
+        :param args: list of all arguments for the call.
+        :type args: list of :py:class:`psyclone.psyir.nodes.Node`
+
+        :raises TypeError: if there is a symbol with the \
+            specified name defined that is not a RoutineSymbol.
+        '''
+        if name in program.symbol_table:
+            routine_symbol = program.symbol_table.lookup(name)
+            if not isinstance(routine_symbol, RoutineSymbol):
+                raise TypeError(
+                    "Error when adding call: Routine '{0}' is "
+                    "a symbol of type '{1}', not a 'RoutineSymbol'."
+                    .format(name, type(routine_symbol).__name__))
+        else:
+            routine_symbol = RoutineSymbol(name)
+            program.symbol_table.add(routine_symbol)
+        call = Call.create(routine_symbol, args)
+        program.addchild(call)
+
+    # -------------------------------------------------------------------------
+    @staticmethod
+    def create_read_in_code(program, psy_data, input_list, output_list,
+                            postfix):
+        '''This function creates the code that reads in the NetCDF file
```

Comment:
Sphinx issues a warning about the formatting of this docstring:

    psyclone.domain.common.extract_driver_creator.ExtractDriverCreator.create_read_in_code:5: WARNING: Block quote ends without a blank line; unexpected unindent.

Reply:
I was about to write 'works for me', then I realised it's in the reference guide. TBH, there are way too many errors triggered there, 'my' errors literally scrolled off screen :(
But very good catch, fixed it.